org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/TopologicalSorter.java

```diff
@@ -0,0 +1,92 @@
+/**
+ * Copyright (c) 2016, 2020 TypeFox GmbH (http://www.typefox.io) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.xtext.ide.server;
+
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.xtext.resource.impl.ProjectDescription;
+import org.eclipse.xtext.util.IAcceptor;
+import org.eclipse.xtext.xbase.lib.IterableExtensions;
+import org.eclipse.xtext.xbase.lib.Procedures.Procedure1;
+
+/**
+ * @author Jan Koehnlein - Initial contribution and API
+ * @since 2.11
+ */
+public class TopologicalSorter {
+	protected static class Entry {
+		private final ProjectDescription description;
+
+		private boolean marked;
+
+		private boolean cyclic;
+
+		@Override
+		public String toString() {
+			return description.getName();
+		}
+
+		public Entry(ProjectDescription description) {
+			this.description = description;
+		}
+	}
+
+	private LinkedHashSet<ProjectDescription> result;
+
+	private Map<String, TopologicalSorter.Entry> name2entry;
+
+	private IAcceptor<ProjectDescription> cyclicAcceptor;
+
+	public List<ProjectDescription> sortByDependencies(Iterable<ProjectDescription> descriptions,
+			Procedure1<? super ProjectDescription> cyclicAcceptor) {
+		this.cyclicAcceptor = new IAcceptor<ProjectDescription>() {
```

Comment:
needed

Reply:
this.cyclicAcceptor = cyclicAcceptor::apply